### PR TITLE
Jarrell Section 7 Basics 11

### PR DESCRIPTION
**Pull request by rocktheship**
Please fill out the survey before submitting the pull request. Thanks!

🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

How many hours did you spend on this assignment?
4-5 hours

Please fill in one error and/or error message you received while working on this assignment.
I create a function that generates a random dice number, and assigned a global variable "rollDiceOne" and executed the function in the global variable level. This caused the "rollDiceOne" and "rollDiceTwo" for players 1 and 2 to be the same numbers.

What part of the assignment did you spend the most time on?
Thinking about how to refactor my code into smaller functions. 

Comfort Level (1-5):
3

Completeness Level (1-5):
4

What did you think of this deliverable?
This was a challenging deliverable.

Is there anything in this code that you feel pleased about?
Finally solving it.

What's one aspect of your code you would like specific, elaborate feedback on?

### CODE REVIEW

File: script.js [hunk 1]
```diff
@@ -1,4 +1,131 @@
...
+};
+
 var main = function (input) {
-  var myOutputValue = 'hello world';
+  // Check if the game mode is "roll the dice".
+  if (gameMode == "roll the dice") {
+    diceRollOne = rollDice();
+    diceRollTwo = rollDice();
+    // Print the welcome message for the player and his/her dice rolls
+    var myOutputValue =
```
Comment (leechuanxin) [inline script.js:73]: I recommend not creating variables with the `var` keyword in control flows like `if` blocks. Variable creation is sensitive in JavaScript. 

Let's assume we are in the "select the dice order" game mode now, as specified in [line 93 to line 129](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L93-L129). Because the `myOutputValue` variable is created only in the "roll the dice" game mode in [line 73](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L73), it will not "exist" per what you have intended in the "select the dice order" game mode.

Your code still works, but it may have certain unintended behaviours. This is because you may have accidentally created the `myOutputValue` on a global scope in lines [100](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L100) and [117](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L117)! Let me explain it this way:
1. In "select the dice order" game mode, [line 73](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L73) with the `var myOutputValue = ...` line is not executed, because the game mode isn't "roll the dice". Thus, we do not have `myOutputValue` scoped to the `main` function.
2. Our intended behaviour on [line 100](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L100) is probably to re-assign values to the `myOutputValue` we thought we have created at [line 73](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L73), right? Instead, JavaScript is going to think: we do not have a variable created called `myOutputValue` now that we are in the "select the dice order" game mode instead of the "roll the dice" game mode. Instead, let me search in the "parent" scope for such a variable.
3. The immediate "parent" scope for the `main` function is the global scope. JavaScript's priority now is to search the global scope for a variable named `myOutputValue`, and then try to re-assign the strings to such a variable in the global scope. You have declared your global variables [between lines 1 and 22](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L1-L22), and JavaScript will notice you do not yet have a `myOutputValue` global variable.
4. Since JavaScript notices that you do not have a `myOutputValue` global variable created, it is going to think, 'Ah, we don't have it. Let me just create one for the developer instead.' Then, _under the hood_, it is going to behave as if you have decided to type in, for instance (using the example at [line 100](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L100)): `var myOutputValue = "You have chosen Dice " + input + " first. Your number is " + firstPlayerSelection + ". <br> It is now Player " + currentPlayerNumber + "'s turn.";` as a global variable.

More succinctly, this strange behaviour is better explained here in a [StackOverflow comment](https://stackoverflow.com/questions/2485423/is-using-var-to-declare-variables-optional/2485439#2485439), as well as [reviewing this Basics lesson about control flows](https://basics.rocketacademy.co/6-conditional-logic/6.1-intro-to-logic).

TL;DR: How do we resolve this? We can do so by just creating the `myOutputValue` value just before we begin writing the `if` and `else if` blocks. We can create it at around [line 68](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L68) (just before the `if` block on line 69) as an empty string, like so: `var myOutputValue = "";`. Then, we do not need to create `myOutputValue` on [line 73](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L73) anymore. [Line 73](https://github.com/rocketacademy/basics-beat-that/blob/cc7cedc8f2744bef5eb8c1589dacd871b6b0e10a/script.js#L73) needs only be a re-assignment of `myOutputValue`, like so:
```
myOutputValue =
  "Welcome to the Beat That game Player " +
  currentPlayerNumber +
  ". <br> You have rolled " +
  diceRollOne +
  " and " +
  diceRollTwo +
  ". <br> Choose the order of the dice by entering '1' or '2'";
```

Notice the lack of the `var` keyword in the code example above?

**Review by leechuanxin — COMMENTED** [review]
Hey Jarrell,

Most of your logic seems fine, so the key takeaway from my comments below is to be aware of creating variables (using the `var` keyword) in control flows like `if` blocks. The TL;DR explanation is a combination of [this StackOverflow comment](https://stackoverflow.com/questions/2485423/is-using-var-to-declare-variables-optional/2485439#2485439), and [reviewing control flows in this Basics lesson](https://basics.rocketacademy.co/6-conditional-logic/6.1-intro-to-logic). 

The combination of these 2 concepts will help you understand why it is unwise to create variables in control flows, and then try to use or re-assign such variables outside of their control flows or `if` blocks.

Other than that, I am glad you have started using and modularising your logic into functions! You can further organise or shorten the `main` function further by breaking up the game states into their own functions. You can review this [Refactoring Code video](https://basics.rocketacademy.co/8-managing-state-and-input-validation/8.2-program-state-for-game-modes#refactoring-code) here on how you can achieve it.

All the best for the [Blackjack final project](https://basics.rocketacademy.co/projects/project-3-blackjack)!
